TESTING.MD

```diff
@@ -0,0 +1,116 @@
+---
+weight: 121
+toc: true
+title: Tests during development
+menu:
+    docs:
+        parent: contributing
+lead: ""
+lastmod: "2023-09-10T08:48:57+00:00"
+images: []
+draft: false
+description: Understanding and running tests when developing Prometheus-Operator
+date: "2023-09-10T00:00:00+00:00"
+---
+
+When contributing code to Prometheus-Operator, you'll notice that every Pull Request will run against an extensive test suite. Among an extensive list of benefits that tests brings to the Project's overall health and reliability, it can be the reviewer's and contributors's best friend during development:
+
+* Test cases serve as documentation, providing insights into the expected behavior of the software.
+* Testing helps prevent regressions by verifying that new changes don't break existing functionality.
+* Running tests locally accelerate the feedback loop, removing the dependency that contributors might have on CI when working on a Pull Request.
+
+This document will focus on teaching you about the different test suites that we currently have and how to run different scenarios to help your development experience!
+
+# Test suites
+
+## Unit tests
+
+Unit tests are used to test particular code snippets in isolation. They are your best ally when looking for quick feedback loops in a particular function. 
+
+Imagine you're working on a PR that adds a new field to the Prometheus CRD and you want to test if your change is reflected to the statefulset. Instead of creating a full Kubernetes cluster, installing all the CRDs, running the Prometheus-Operator, deploying a Prometheus resource and finally check if your change made it to the live object, you could simply write or extend a unit test for the statefulset generation.
+
+Here is an example test that checks if labels are present in the statefulset.
+
+https://github.com/prometheus-operator/prometheus-operator/blob/aeceb0b4fadc8307a44dc55afdceca0bea50bbb0/pkg/prometheus/server/statefulset_test.go#L140-L167
+
+Unit tests can be run with:
+
+```
+make test-unit
+```
+
+They can also be run for particular packages:
+
+```
+go test ./pkg/prometheus/server
+```
+
+Or even particular functions:
+
+```
+go test -run ^TestPodLabelsAnnotations$ ./pkg/prometheus/server
+```
+
+### Testing multiline string comparison - Golden files
+
+Golden files are plain text files used to assist assertion of long strings. In Prometheus-Operator they are often often generated by a function to be used as part of a ConfigMap or Secret.
```

Comment:
IMO this part is a bit trick to understand, looks like we're using golden-files to store data in a config map or secret.

Maybe we can have something like that.

```suggestion
Golden files are plaintext documents designed to facilitate the validation of lengthy strings. They come in handy when, for instance, you need to test a Prometheus Config that's generated using GoLang structures. You can marshal this configuration into YAML and then compare it against a static reference to ensure a match. Golden files offer an elegant solution to this challenge, sparing you the need to hard-code the static configuration directly into your test code.
```

Reply:
Good suggestion, did you copy that sentence from somewhere? Looks too elegant 😛

Reply:
I made a initial idea by myself and than get some help from Grammarly and ChatGPT 😛